torcharrow/test/integration/test_criteo.py

```diff
@@ -0,0 +1,176 @@
+# TODO: this is not a general purpose JaggedTensor conversion -- it leverages the fact that in Criteo preproc, each array is single element
+# TODO: implement general purpose jagged sparse tensor conversion in native C++
+class _CriteoJaggedTensorConversion(tap.ITorchConversion):
+    def to_tensor(self, df: ta.IDataFrame):
+        # TODO: Implement df.size(), similar to Pandas
+        num_arrays = len(df) * len(df.columns)
+
+        keys: List[str] = df.columns
+        lengths: torch.Tensor = torch.ones(
+            (num_arrays, ), dtype=torch.int32,
+        )
+        length_per_key: List[int] = len(df.columns) * [len(df)]
+
+        # Generate values
+        tensors = df.to_tensor()
+        # Stack them into BatchSize * NumFields Tensor
+        packed_tensor = torch.cat([column_tensor.values.values.values.unsqueeze(0).T for column_tensor in tensors], dim=1)
+        # Flatten the tensors into one-dimension with columnar ordering
+        values = packed_tensor.transpose(1, 0).reshape(-1)
+
+        return (keys, values, lengths, length_per_key)
+
+
+class CriteoIntegrationTest(unittest.TestCase):
+    NUM_ROWS = 128
+    RAW_ROWS = None
+    TEMPORARY_PARQUETY_FILE = "_test_criteo.parquet"
+
+    def setUp(self):
+        # Generate some random data
+        random.seed(42)
+
+        rows = []
+        for i in range(type(self).NUM_ROWS):
+            label = 1 if random.randrange(100) < 4 else 0 
+            dense_features_struct = tuple(
+                    random.randrange(1000) if random.randrange(100) < 80 else None for j in range(INT_FEATURE_COUNT)
+                )
+            sparse_features_struct = tuple(
+                    random.randrange(-2 ** 31, 2 ** 31) if random.randrange(100) < 98 else None for j in range(CAT_FEATURE_COUNT)
+                )
+
+            rows.append((label, dense_features_struct, sparse_features_struct))
+
+        type(self).RAW_ROWS = rows
+        df = ta.DataFrame(rows, dtype=DTYPE)
+
+        pq.write_table(df.to_arrow(), type(self).TEMPORARY_PARQUETY_FILE)
+    
+    def tearDown(self) -> None:
+        os.remove(type(self).TEMPORARY_PARQUETY_FILE)
+
+    @staticmethod
+    def preproc(df: ta.IDataFrame) -> ta.IDataFrame:
+        # 1. fill null values
+        df["dense_features"] = df["dense_features"].fill_null(0)
+        df["sparse_features"] = df["sparse_features"].fill_null(0)
+
+        # 2. apply log(x+3) on dense features
+        df["dense_features"] = (df["dense_features"] + 3).log()
+
+        # 3. Pack each categorical feature as an single element array
+        sparse_features = df["sparse_features"]
+        for field in DEFAULT_CAT_NAMES:
+            sparse_features[field] = functional.array_constructor(sparse_features[field])
+
+        # FIXME: we shouldn't need to "put back the struct column"
+        df["sparse_features"] = sparse_features
+
+        df["label"] = df["label"].cast(dt.int32)
```

Comment:
cast will also be improved via https://github.com/facebookresearch/torcharrow/issues/143